source/material_model/averaging.cc

```diff
@@ -0,0 +1,412 @@
+/*
+  Copyright (C) 2011 - 2015 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+#include <deal.II/base/std_cxx11/array.h>
+#include <aspect/material_model/averaging.h>
+#include <deal.II/base/parameter_handler.h>
+#include <utility>
+#include <limits>
+
+using namespace dealii;
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    // Do the requested averaging operation for one array.
+    template <int dim>
+    void
+    Averaging<dim>::average (const std::string &averaging_operation,
+                             const std::vector<Point<dim> >    &position,
+                             std::vector<double>           &values_out) const
+    {
+      // if an output field has not been filled (because it was
+      // not requested), then simply do nothing -- no harm no foul
+      if (values_out.size() == 0)
+        return;
+
+      const unsigned int N = values_out.size();
+      std::string averaging_operation_lower = averaging_operation;
+
+      // switch to lower case, so the input becomes independent of case
+      std::transform(averaging_operation_lower.begin(), averaging_operation_lower.end(), averaging_operation_lower.begin(), ::tolower);
+
+      // perform the requested averaging
+      if (averaging_operation_lower == "none")
+        {
+        }
+      else if (averaging_operation_lower == "arithmetic average")
+        {
+          double sum = 0;
```

Comment:
I would suggest to drop the averaging operations that are already implemented separately. This only duplicates code, but does not really add functionality that wasn't already there.